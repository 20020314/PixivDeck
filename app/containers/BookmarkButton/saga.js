// @flow
import { call, put, takeEvery, type IOEffect } from 'redux-saga/effects'
import { get } from '../Api/sagas'
import * as api from 'containers/Api/sagas'
import * as Actions from './constants'
import * as actions from './actions'
import type { Restrict } from './types'

type Props = {
  id: number,
  restrict: Restrict,
}

export function* bookmark({ id, restrict }: Props): Generator<*, void, void> {
  try {
    yield call(api.post, '/v2/illust/bookmark/add', { illustId: id, restrict })

    yield put(actions.addBookmarkSuccess(id, restrict))
  } catch (err) {
    yield put(actions.addBookmarkFailer(id, err))
  }
}

export function* deleteTask({ id }: Props): Generator<*, void, void> {
  try {
    yield call(api.post, '/v1/illust/bookmark/delete', { illustId: id })
    yield put(actions.deleteBookmarkSuccess(id))
  } catch (err) {
    yield put(actions.deleteBookmarkFailer(id, err))
  }
}

function* success({ id }: Props) {
  try {
    yield call(get, `/v1/illust/detail?illust_id=${id}`, true)
  } catch (err) {}
}

function* root(): Generator<IOEffect, void, *> {
  yield takeEvery(Actions.ADD_BOOKMARK_REQUEST, bookmark)
  yield takeEvery(Actions.DELETE_BOOKMARK_REQUEST, deleteTask)
  yield takeEvery(
    [Actions.ADD_BOOKMARK_SUCCESS, Actions.DELETE_BOOKMARK_SUCCESS],
    success
  )
}

export default root
