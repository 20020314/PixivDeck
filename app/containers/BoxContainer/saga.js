// @flow
import { shell } from 'electron'
import { Actions } from './actionTypes'
import { call, takeEvery, type IOEffect } from 'redux-saga/effects'

function* openPixiv({ id }: { id: number }): Generator<IOEffect, void, *> {
  yield call(
    shell.openExternal,
    `http://www.pixiv.net/member_illust.php?mode=medium&illust_id=${id}`
  )
}

export default function* root(): Generator<IOEffect, void, *> {
  yield takeEvery(Actions.OPEN_PIXIV, openPixiv)
}
