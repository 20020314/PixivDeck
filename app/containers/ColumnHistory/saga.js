// @flow
import { addColumn } from 'containers/ColumnManager/actions'
import * as Actions from './constants'
import * as actions from './actions'
import { put, takeEvery } from 'redux-saga/effects'
import { OPEN_ILLUST_VIEWER } from '../IllustPreview/constants'
import { OPEN_MANGA_PREVIEW } from '../MangaPreview/constants'

export function* addHistoryColumn(): Generator<*, void, void> {
  yield put(addColumn('HISTORY', { columnId: 'history', type: 'HISTORY' }))
}

type Action = { id: number }

export function* addHistory({ id }: Action): Generator<*, void, void> {
  yield put(actions.addHistory(id))
}

export default function* root(): Generator<*, void, void> {
  yield takeEvery(Actions.ADD_COLUMN_HISTORY, addHistoryColumn)

  yield takeEvery(OPEN_ILLUST_VIEWER, addHistory)
  yield takeEvery(OPEN_MANGA_PREVIEW, addHistory)
}
