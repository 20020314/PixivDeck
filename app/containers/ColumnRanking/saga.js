// @flow
import union from 'lodash/union'
import { addColumn } from 'containers/ColumnManager/actions'
import { getRequest } from '../../api/client'
import * as Actions from './constants'
import * as actions from './actions'
import type { Mode } from './reducer'
import { makeSelectColumn, makeSelectModes } from './selectors'
import { put, select, call, takeEvery } from 'redux-saga/effects'

function* addRakingColumn({ mode }: { mode: Mode }) {
  const modes: Array<?Mode> = yield select(makeSelectModes())
  if (modes.every(v => v !== mode)) {
    yield put(actions.addRankingColumnSuccess(mode))
  }

  yield put(addColumn(`ranking-${mode}`, { columnId: mode, type: 'RANKING' }))
}

type Props = { id: Mode }

function* fetchRanking(props: Props) {
  const { id } = props
  const { illustIds } = yield select(makeSelectColumn(), props)

  try {
    const response = yield call(getRequest, `/v1/illust/ranking?mode=${id}`)
    const { result } = response

    yield put(actions.setNextUrl(id, result.nextUrl))

    const nextIds = union(illustIds, result.illusts)
    yield put(actions.fetchRankingSuccess(id, response, nextIds))
  } catch (err) {
    yield put(actions.fetchRankingFailre(id))
  }
}

function* fetchNextRanking(props: Props) {
  const { id } = props
  const { illustIds, nextUrl } = yield select(makeSelectColumn(), props)

  try {
    if (!nextUrl) {
      return
    }

    const response = yield call(getRequest, nextUrl)
    const { result } = response

    yield put(actions.setNextUrl(id, result.nextUrl))

    const nextIds = union(illustIds, result.illusts)
    yield put(actions.fetchRankingSuccess(id, response, nextIds))
  } catch (err) {
    yield put(actions.fetchNextRankingFailre(id))
  }
}

export default function* root(): Generator<*, void, void> {
  yield takeEvery(Actions.ADD_RANKING_COLUMN, addRakingColumn)
  yield takeEvery(Actions.FETCH_RANKING, fetchRanking)
  yield takeEvery(Actions.FETCH_NEXT_RANKING, fetchNextRanking)
}
