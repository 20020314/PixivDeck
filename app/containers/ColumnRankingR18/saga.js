// @flow
import { addTable } from 'containers/ColumnManager/actions'
import * as fetchColumn from '../Column/sagas'
import * as Actions from './constants'
import * as actions from './actions'
import type { R18Mode } from './reducer'
import { makeSelectColumn, makeSelectModes } from './selectors'
import { put, select, call, takeEvery } from 'redux-saga/effects'

type Action = {
  +id: R18Mode,
}

export function* addColumn({ id }: Action): Generator<*, void, *> {
  const modes: Array<?R18Mode> = yield select(makeSelectModes())

  if (modes.every(v => v !== id)) {
    yield put(actions.addColumnSuccess(id))
  }

  yield put(
    addTable(`ranking-r18-${id}`, { columnId: id, type: 'RANKING_R18' })
  )
}

function* fetchRanking(action: Action) {
  const { ids } = yield select(makeSelectColumn(), action)
  const endpoint = `/v1/illust/ranking?mode=${action.id}`
  yield call(fetchColumn.fetchColumn, endpoint, action.id, actions, ids)
}

function* fetchNextRanking18(action: Action) {
  const { ids, nextUrl } = yield select(makeSelectColumn(), action)
  yield call(fetchColumn.fetchColumn, nextUrl, action.id, actions, ids)
}

export default function* root(): Generator<*, void, void> {
  yield takeEvery(Actions.ADD_COLUMN, addColumn)
  yield takeEvery(Actions.FETCH, fetchRanking)
  yield takeEvery(Actions.FETCH_NEXT, fetchNextRanking18)
}
