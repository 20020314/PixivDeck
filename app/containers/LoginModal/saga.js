// @flow
// eslint-disable-next-line import/order
import { select, put, call, takeEvery } from 'redux-saga/effects'
import { fetchAuth } from 'services/api'
import { openModal, closeModal } from '../ModalManeger/actions'
import {
  loginFailure,
  endLoading,
  clearError,
  setAuth,
  setAccount,
} from './actions'
import * as Actions from './constants'
import { makeSelectInfo } from './selectors'

export function* getToken(): Generator<*, string, *> {
  const info = yield select(makeSelectInfo())
  // TODO: username & passwordがなければLogin Pageを開く
  const { accessToken } = yield call(fetchAuth, info)
  return accessToken
}

function* authorize({ username, password }): Generator<*, void, *> {
  // エラーを非表示
  yield put(clearError())
  try {
    const { user: account } = yield call(fetchAuth, { username, password })
    yield put(setAuth(username, password))
    yield put(setAccount(account))
    yield put(closeModal())
  } catch (err) {
    yield put(loginFailure())
  } finally {
    yield put(endLoading())
  }
}

function* logout() {
  yield put(endLoading())
  // ログインモーダルを表示
  yield put(openModal('Login'))
}

function* root(): Generator<*, void, void> {
  yield takeEvery(Actions.LOGIN_REQUEST, authorize)
  yield takeEvery(Actions.LOGOUT, logout)
}

export default root
