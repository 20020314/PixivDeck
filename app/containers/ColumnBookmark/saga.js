// @flow
import union from 'lodash/union'
import { addColumn } from 'containers/ColumnManager/actions'
import { makeSelectInfo } from 'containers/LoginModal/selectors'
import { getToken } from 'containers/LoginModal/saga'
import { getRequest, fetchAuth } from '../../api/client'
import * as Actions from './constants'
import * as actions from './actions'
import type { ColumnId } from './reducer'
import { makeSelectColumn, makeSelectIds } from './selectors'
import { put, select, call, takeEvery } from 'redux-saga/effects'

function* addBookmarkColumn({ id }: { id: ColumnId }) {
  const ids: Array<?ColumnId> = yield select(makeSelectIds())
  if (ids.every(v => v !== id)) {
    yield put(actions.addBookmarkColumnSuccess(id))
  }

  yield put(addColumn(`bookmark-${id}`, { columnId: id, type: 'BOOKMARK' }))
}

type Props = { id: ColumnId }

function* fetchBookmark(props: Props) {
  const { id } = props
  const { illustIds } = yield select(makeSelectColumn(), props)

  try {
    const info = yield select(makeSelectInfo())
    // TODO
    const { accessToken, user: { id: userId } } = yield call(fetchAuth, info)

    const response = yield call(
      getRequest,
      `/v1/user/bookmarks/illust?user_id=${userId}&restrict=${id}`,
      null,
      accessToken
    )
    const { result } = response

    yield put(actions.setNextUrl(id, result.nextUrl))

    const nextIds = union(result.illusts, illustIds)
    yield put(actions.fetchBookmarkSuccess(id, response, nextIds))
  } catch (err) {
    yield put(actions.fetchBookmarkFailre(id))
  }
}

function* fetchNextBookmark(props: Props) {
  const { id } = props
  const { illustIds, nextUrl } = yield select(makeSelectColumn(), props)

  try {
    if (!nextUrl) {
      return
    }

    const accessToken = yield call(getToken)

    const response = yield call(getRequest, nextUrl, null, accessToken)
    const { result } = response

    yield put(actions.setNextUrl(id, result.nextUrl))

    const nextIds = union(illustIds, result.illusts)
    yield put(actions.fetchBookmarkSuccess(id, response, nextIds))
  } catch (err) {
    yield put(actions.fetchNextBookmarkFailre(id))
  }
}

export default function* root(): Generator<*, void, void> {
  yield takeEvery(Actions.ADD_BOOKMARK_COLUMN, addBookmarkColumn)
  yield takeEvery(Actions.FETCH_BOOKMARK, fetchBookmark)
  yield takeEvery(Actions.FETCH_NEXT_BOOKMARK, fetchNextBookmark)
}
