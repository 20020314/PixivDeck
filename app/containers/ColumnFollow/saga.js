// @flow
import { union, difference } from 'lodash'
import { delay } from 'redux-saga'
import { put, select, call, takeEvery } from 'redux-saga/effects'
import { addColumn } from 'containers/ColumnManager/actions'
import { makeSelectInfo } from 'containers/LoginModal/selectors'
import { getToken } from 'containers/LoginModal/saga'
import { getRequest, fetchAuth } from 'services/api'
import * as Actions from './constants'
import * as actions from './actions'
import type { ColumnId } from './reducer'
import { makeSelectColumn, makeSelectIds } from './selectors'
import * as selectors from './selectors'
import { addNotifyWithIllust } from '../Notify/actions'
import { FOLLOW_SUCCESS } from '../FollowButton/constants'

function* addFollowColumn({ id }: { id: ColumnId }) {
  const ids: Array<?ColumnId> = yield select(makeSelectIds())
  if (ids.every(v => v !== id)) {
    yield put(actions.addFollowColumnSuccess(id))
  }

  yield put(addColumn(`follow-${id}`, { columnId: id, type: 'FOLLOW' }))
}

type Action = { id: ColumnId }

function createEndpoint(userId, restrict) {
  return `/v2/illust/follow?user_id=${userId}&restrict=${restrict}`
}

function* fetchFollow(action: Action): Generator<*, void, *> {
  const { id } = action
  try {
    const { illustIds } = yield select(makeSelectColumn(), action)

    // TODO
    const info = yield select(makeSelectInfo())
    const { accessToken, user: { id: userId } } = yield call(fetchAuth, info)

    const response = yield call(
      getRequest,
      `/v2/illust/follow?user_id=${userId}&restrict=${id}`,
      null,
      accessToken
    )
    const { result } = response

    yield put(actions.setNextUrl(id, result.nextUrl))

    const nextIds = union(illustIds, result.illusts)

    if (illustIds.length > 0) {
      const diffIllusts = difference(nextIds, illustIds)
      for (const id of diffIllusts) {
        yield put(addNotifyWithIllust('新着イラスト', id))
      }
    }

    yield put(actions.fetchFollowSuccess(id, response, nextIds))
  } catch (err) {
    yield put(actions.fetchFollowFailre(id, err))
  }
}

function* fetchNextFollow(action: Action) {
  const { id } = action
  try {
    const { illustIds, nextUrl } = yield select(makeSelectColumn(), action)

    if (!nextUrl) {
      return
    }

    const accessToken = yield call(getToken)

    const response = yield call(getRequest, nextUrl, null, accessToken)
    const { result } = response

    yield put(actions.setNextUrl(id, result.nextUrl))

    const nextIds = union(illustIds, result.illusts)
    yield put(actions.fetchNextFollowSuccess(id, response, nextIds))
  } catch (err) {
    yield put(actions.fetchNextFollowFailre(id, err))
  }
}

function* fetchNew(action: Action): Generator<*, void, *> {
  try {
    const { illustIds } = yield select(selectors.makeSelectColumn(), action)

    const info = yield select(makeSelectInfo())
    const { accessToken, user: { id: userId } } = yield call(fetchAuth, info)

    const endpoint = createEndpoint(userId, action.id)

    const response = yield call(getRequest, endpoint, null, accessToken)
    const { result } = response

    const nextIds = union(result.illusts, illustIds)
    yield put(actions.fetchNewSuccess(action.id, response, nextIds))
  } catch (err) {
    yield put(actions.fetchNewFailre(action.id, err))
  }
}

// TODO キャンセル
function* fetchNewWatch(action: Action) {
  try {
    while (true) {
      const { interval } = yield select(selectors.makeSelectColumn(), action)
      yield call(fetchNew, action)
      yield call(delay, interval)
    }
  } catch (err) {
    // TODO エラーハンドリング
    console.log(err)
  }
}

export default function* root(): Generator<*, void, void> {
  yield takeEvery(Actions.ADD_FOLLOW_COLUMN, addFollowColumn)
  yield takeEvery(Actions.FETCH_FOLLOW, fetchFollow)
  yield takeEvery(Actions.FETCH_NEXT_FOLLOW, fetchNextFollow)

  yield takeEvery(Actions.FETCH_FOLLOW_SUCCESS, fetchNewWatch)
  yield takeEvery(FOLLOW_SUCCESS, function*({ restrict }) {
    yield call(fetchFollow, { id: restrict })
  })
}
