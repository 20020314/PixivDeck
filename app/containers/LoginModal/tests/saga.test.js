// @flow
import { call } from 'redux-saga/effects'
import * as sagas from '../saga'
import { fetchAuth } from 'services/api'

test('getToken', () => {
  const gen = sagas.getToken()

  let next = gen.next()
  expect(next.value).toMatchSnapshot()

  const fakeUserInfo = { username: 'fake-username', password: 'fake-pass' }
  next = gen.next(fakeUserInfo)
  expect(next.value).toEqual(call(fetchAuth, fakeUserInfo))

  expect(gen.return('fake-token')).toMatchSnapshot()
})
