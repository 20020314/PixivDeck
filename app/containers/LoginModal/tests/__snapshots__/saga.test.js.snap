// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getToken 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getToken 2`] = `
Object {
  "done": true,
  "value": "fake-token",
}
`;
